### PR TITLE
Core: add key_metadata to ManifestFile spec

### PR DESCRIPTION
**Pull request by jackye1995**
@flyrain @rdblue @ggershinsky 

first step based on draft #2520, simply add `keyMetadata` to the interface and implementations, pass in `null` for any caller.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/GenericManifestFile.java
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/GenericManifestFile.java:418]: Given we're adding a number of encryption related PRs, most of which have very sensitive data in them (encryption keys), would it make sense to make this into a utility function, such as `EncryptionUtils.toRedactedString(byte[] value)`?

We could have a unified way of redacting, which could also reduce the null checks in the code.

It would give people the option of possibly customizing the way in which they redact keys (say, in a fork or something or via an override) that redacts in the way that allows SREs to better assist customers - for example, I can imagine that there _might_ be utility in showing the first 2 or 4 bytes or something so as to be able to check that the key is definitively not the same.

Not sure how much benefit people would have in custom redaction for stringifying, but throwing that out there as I'm curious to hear if there is any utility in not entirely redacting, from a usability / debugging standpoint.

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/GenericManifestFile.java:418]: I would actually be more against using this approach to support redact, for 2 reasons:

1. this is only for `toString`. If you really want to see the details of the key metadata you can read through the manifest table, and that should show the information fully if you can decrypt an encrypted manifest list. I think that provides better access control. Changing `toString` feels to me like opening a backdoor.
2. if we can encrypt a manifest, we should be able to support redact configuration natively through encryption manager as a part of data masking. When we can do that, you don't need a forked version to use a custom static method for this purpose.

File: core/src/main/java/org/apache/iceberg/GenericManifestFile.java
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/GenericManifestFile.java:131]: Do we need a null guard here?

File: core/src/main/java/org/apache/iceberg/GenericManifestFile.java
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/GenericManifestFile.java:418]: So maybe we should just drop this from toString? I'm fine with that

**Review by RussellSpitzer — APPROVED** [review]
This looks good to me, do we have anyone else we are waiting on to get this merged? @aokolnychyi ?

**Review by flyrain — APPROVED** [review]
+1, thanks for the PR.

File: core/src/main/java/org/apache/iceberg/GenericManifestFile.java
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/GenericManifestFile.java:131]: There is already a null guard in `ByteBuffers`

**Comment by jackye1995** [comment]
@RussellSpitzer we are not waiting for anyone to merge, but I thought for any format v2 change @rdblue should take a look and approve. If it's not necessary we will merge to unblock other subsequent changes.

File: api/src/test/java/org/apache/iceberg/TestHelpers.java
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/TestHelpers.java:354]: Is this used in tests? We could just allow this to use the default implementation, unless we added tests that exercise it.

File: core/src/main/java/org/apache/iceberg/GenericManifestFile.java
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/GenericManifestFile.java:418]: I'm also fine dropping this from `toString`. I think the main concern is that we don't want to dump binary key metadata as a string. I'm fine if this shows nothing at all.

**Review by rdblue — APPROVED** [review]

**Pull request merged**
